File: Backend/fastapi_app/app/api/v1/endpoints/resources.py
"""
Resource endpoints:
- GET /resources with provider/status filters and pagination
"""
from __future__ import annotations

from typing import List, Optional

from fastapi import APIRouter, Depends, Query
from sqlalchemy import Select, select
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.security import require_jwt_claims
from app.db.session import get_session
from app.models.schemas import ResourceOut
from app.models.tables import Resource

router = APIRouter(tags=["Resources"])


# PUBLIC_INTERFACE
@router.get(
    "/resources",
    response_model=List[ResourceOut],
    summary="List all resources across clouds",
    description="Supports optional filtering by provider and status. Pagination via page & pageSize (defaults: 1, 50).",
    responses={200: {"description": "List of resources"}, 401: {"description": "Unauthorized"}},
)
async def list_resources(
    provider: Optional[str] = Query(default=None, description="Cloud provider (AWS|Azure|GCP)"),
    status: Optional[str] = Query(default=None, description="Resource status"),
    page: int = Query(default=1, ge=1, description="Page number (1-based)"),
    page_size: int = Query(default=50, ge=1, le=200, alias="pageSize", description="Page size"),
    _: dict = Depends(require_jwt_claims),
    session: AsyncSession = Depends(get_session),
) -> List[ResourceOut]:
    """List resources with filters and pagination."""
    stmt: Select = select(Resource)
    if provider:
        stmt = stmt.where(Resource.provider == provider)
    if status:
        stmt = stmt.where(Resource.status == status)

    stmt = stmt.order_by(Resource.created_at.desc(), Resource.name.asc())
    offset = (page - 1) * page_size
    stmt = stmt.offset(offset).limit(page_size)

    res = await session.execute(stmt)
    resources = res.scalars().all()
    return [ResourceOut.model_validate(r) for r in resources]

File: Backend/legacy_express/src/services/resources.js
'use strict';

const { query } = require('../db');

function mapResourceRow(row) {
  return {
    id: row.id,
    provider: row.provider,
    type: row.type,
    name: row.name,
    tags: row.tags || {},
    cost: Number(row.cost || 0),
    status: row.status,
    createdAt: row.created_at ? new Date(row.created_at).toISOString() : null,
  };
}

// PUBLIC_INTERFACE
async function listResources(filters = {}) {
  const values = [];
  const where = [];
  if (filters.provider) {
    where.push(`provider = $${values.length + 1}`);
    values.push(filters.provider);
  }
  if (filters.status) {
    where.push(`status = $${values.length + 1}`);
    values.push(filters.status);
  }

  const sql = `SELECT id, provider, type, name, tags, cost, status, created_at
               FROM resources
               ${where.length ? 'WHERE ' + where.join(' AND ') : ''}
               ORDER BY created_at DESC NULLS LAST, name ASC`;
  const result = await query(sql, values);
  return result.rows.map(mapResourceRow);
}

module.exports = {
  listResources,
};
